opal/mca/accelerator/cuda/accelerator_cuda.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2014      Intel, Inc.  All rights reserved.
+ * Copyright (c) 2017-2022 Amazon.com, Inc. or its affiliates.
+ *                         All Rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#ifndef MCA_ACCELERATOR_CUDA_H
+#define MCA_ACCELERATOR_CUDA_H
+
+#include "opal_config.h"
+#include "opal/mca/accelerator/accelerator.h"
+#include "opal/mca/threads/mutex.h"
+#include <cuda.h>
+
+BEGIN_C_DECLS
+
+typedef struct {
+    opal_accelerator_base_component_t super;
+} opal_accelerator_cuda_component_t;
+
+/* Structure to hold CUDA function pointers that get dynamically loaded. */
+struct  accelerator_cuda_func_table {
```

Comment:
Does anyone have any objections to just dlopening libcuda and getting rid of the function table (since it will now be guarded behind the CUDA component of the accelerator framework) 

@janjust

Reply:
@open-mpi/cuda

Reply:
So after doing more research into this, we still need to keep the CUDA function table.  The reason is that we want to support the package manager use case.  The package managers want a openMPI library compiled with --with-cuda, and they want to be able to run it on machines that don't have CUDA libraries.  If we want that use case to work, we cannot link cuda into our application at all (b/c then the linker would fail to find the library, and the program wouldn't start).

The way dlopen, dlsym, and dlclose work for loading dynamic libraries at runtime is they return the symbols address, i.e..:
```
handle = dlopen(“foo.so”, RTLD_NOW | RTLD_LOCAL);
(void (*)(void)) foo = (void (*)(void)) dlsym(handle,”the_foo_function”);
```

If someone wants to call the the_foo_function, they must have access to the function pointer returned from dlsym (i.e.. the function table).

Reply:
That seems motivation to have the cuda component built as a dso instead as part of libopen-pal, and you get the same thing.

Reply:
I'm going to state this stronger.  I don't see any reason for the function table.  There's no reason that cuda should be different than OFI or UCX or any of the other multitude of components with external requirements.